monai/utils/deprecated.py

```diff
@@ -0,0 +1,160 @@
+# Copyright 2020 - 2021 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import inspect
+import warnings
+from functools import wraps
+from threading import Lock
+from types import FunctionType
+from typing import Optional
+
+from .. import __version__
+
+__all__ = ["deprecated", "deprecated_arg", "DeprecatedError"]
+
+warned_set = set()
+warned_lock = Lock()
+
+
+class DeprecatedError(Exception):
+    pass
+
+
+def warn_deprecated(obj, msg):
+    """
+    Issue the warning message `msg` only once per process for the given object `obj`. When this function is called
+    and `obj` is not in `warned_set`, it is added and the warning issued, if it's already present nothing happens.
+    """
+    if obj not in warned_set:  # ensure warning is issued only once per process
+        warned_set.add(obj)
+        warnings.warn(msg, category=DeprecationWarning, stacklevel=2)
+
+
+def version_leq(lhs, rhs):
+    """Returns True if version `lhs` and earlier or equal to `rhs`."""
+
+    if "+" in lhs:
+        lhs = lhs.split("+", 1)[0]
+
+    if "+" in rhs:
+        rhs = rhs.split("+", 1)[0]
+
+    def _try_cast(val):
+        val = val.strip()
+        try:
+            return int(val)
+        except ValueError:
+            return val
+
+    # parse the version strings in this basic way to avoid needing the `packaging` package
+    lhs = map(_try_cast, lhs.split("."))
+    rhs = map(_try_cast, rhs.split("."))
+
+    return all(l <= r for l, r in zip(lhs, rhs))
+
+
+def deprecated(
+    since: Optional[str] = None, removed: Optional[str] = None, msg_suffix: str = "", version_val=__version__
+):
+    """
+    Marks a function or class as deprecated. If `since` is given this should be a version at or earlier than the
+    current version and states at what version of the definition was marked as deprecated. If `removed` is given
+    this can be any version and marks when the definition was removed. When the decorated definition is called,
+    that is when the function is called or the class instantiated, a warning is issued if `since` is given and
+    the current version is at or later than that given. An exception is instead raised if `removed` is given and
+    the current version is at or later than that, or if neither `since` nor `removed` is provided.
+    """
+
```

Comment:
Could you please help add the doc-string for `Args`? I am confused by `msg_suffix` and `version_val` at first look.

Thanks.